lib/Parser/BGParseManager.cpp

```diff
@@ -249,13 +268,51 @@ HRESULT BGParseManager::GetParseResults(Js::ScriptContext* scriptContextUI, DWOR
     return hr;
 }
 
+// Finds and removes the workitem associated with the provided cookie. If the workitem is processed
+// or not yet processed, the workitem is simply removed and freed. If the workitem is being processed,
+// it is removed from the list and will be freed after the job is processed (with its script source
+// buffer).
+// Returns true when the caller should free the script source buffer. Otherwise, when false is returned,
+// the workitem is responsible for freeing the script source buffer.
+bool BGParseManager::DiscardParseResults(DWORD cookie, void* buffer)
```

Comment:
Is there any way we could simplify the memory management semantics here? Maybe DiscardParseResults could take ownership of the buffer in all cases, and free it either immediately (work item not found, not yet processed, or already done) or deferred (work item in progress).

Reply:
I'm not 100% about the handshake between the host and Chakra on this. This model keeps things as they are until it's absolutely needed.
I'll followup with someone who knows more than me :-) Thanks!!